09_json/willshen8/pkg/puppy/store/storer_test.go

```diff
@@ -0,0 +1,121 @@
+		Message: "Puppy ID can not be found, read operation failed."}
+	UpdateError = puppy.Error{Code: puppy.NonExistentPuppy,
+		Message: "Puppy ID can not be found, update operation failed."}
+	DeleteError = puppy.Error{Code: puppy.NonExistentPuppy,
+		Message: "Puppy ID can not be found, delete operation failed."}
+	WrongValueError = puppy.Error{Code: puppy.ErrorValueFormat,
+		Message: "Unrecongised puppy value."}
+	ErrorCreateID = uint32(0)
+)
+
+type StorerSuite struct {
+	suite.Suite
+	storer Storer
+}
+
+func (s *StorerSuite) TestCreatePuppySuccessfully() {
+	m := assert.New(s.T())
+	createdID, _ := s.storer.CreatePuppy(&Puppy1)
+	readPuppy1, _ := s.storer.ReadPuppy(createdID)
+	m.Equal(&Puppy1, readPuppy1)
+}
+
+func (s *StorerSuite) TestCreatePuppyWithValueLessThanZero() {
+	createError := "Error code 1001: Puppy value can't be less than 0.\n"
+	m := assert.New(s.T())
+	var buf bytes.Buffer
+	out = &buf
+	createdID, error := s.storer.CreatePuppy(&InvalidPuppy)
+	fmt.Fprintln(out, error)
+	actualError := buf.String()
+	m.Equal(createError, actualError)
+	m.Equal(ErrorCreateID, createdID)
+}
+
+func (s *StorerSuite) TestCreatePuppyWithNoValue() {
+	m := assert.New(s.T())
+	_, err := s.storer.CreatePuppy(&NoValuePuppy)
+	m.Equal(&WrongValueError, err)
+}
+
+func (s *StorerSuite) TestCreateNextPuppyWithCorrectID() {
+	m := assert.New(s.T())
+	createID1, creatError1 := s.storer.CreatePuppy(&Puppy2)
+	if createID1 != uint32(1) && creatError1 != nil {
+		panic("Error creating puppy")
+	}
+	// create a second puppy should give an id of 2
+	createdID2, _ := s.storer.CreatePuppy(&Puppy3)
+	m.Equal(uint32(2), createdID2)
+}
+
+func (s *StorerSuite) TestReadPuppySuccessfully() {
+	m := assert.New(s.T())
+	returnedPuppyID4, _ := s.storer.CreatePuppy(&Puppy4)
+	returnPuppy4, _ := s.storer.ReadPuppy(returnedPuppyID4)
+	m.Equal(&Puppy4, returnPuppy4)
+}
+
+func (s *StorerSuite) TestReadNonExistentPuppy() {
+	m := assert.New(s.T())
+	returnPuppy, error := s.storer.ReadPuppy(100)
+	m.Nil(returnPuppy)
+	m.Equal(&ReadError, error)
+}
+
+func (s *StorerSuite) TestUpdatePuppySuccessfully() {
+	returnedPuppyID, _ := s.storer.CreatePuppy(&Puppy5)
+	updateResult, _ := s.storer.UpdatePuppy(returnedPuppyID, &Puppy6)
+	s.Equal(true, updateResult)
+}
+
+func (s *StorerSuite) TestUpdateNonExistentID() {
+	updateResult, error := s.storer.UpdatePuppy(888, &Puppy1)
+	s.Equal(&UpdateError, error)
+	s.Equal(false, updateResult)
+}
+
+func (s *StorerSuite) TestDeletePuppy() {
+	createdPuppyID, _ := s.storer.CreatePuppy(&Puppy7)
+	deleteResult, _ := s.storer.DeletePuppy(createdPuppyID)
+	s.Equal(true, deleteResult)
+	returnPuppy, _ := s.storer.ReadPuppy(createdPuppyID)
+	s.Nil(returnPuppy)
+}
+
+func (s *StorerSuite) TestDeleteNonExistentPuppy() {
+	deleteResult, error := s.storer.DeletePuppy(uint32(10000))
+	s.Equal(false, deleteResult)
+	s.Equal(&DeleteError, error)
+}
+
+func TestStorer(t *testing.T) {
+	suite.Run(t, &StorerSuite{storer: NewMapStore()})
```

Comment:
Same [comments as in lab8](https://github.com/anz-bank/go-course/pull/555#discussion_r308614336) are valid here too IMO

Reply:
I have done this for both labs, please let me know if this is still not quite right.

Reply:
lgtm